experimental/adaptive-card-prompt-sample/javascript_nodejs/adaptiveCardPrompt.js

```diff
@@ -0,0 +1,222 @@
+        this._missingRequiredInputsMessage = settings.missingRequiredInputsMessage || 'The following inputs are required';
+        this._attemptsBeforeCardRedisplayed = settings.attemptsBeforeCardRedisplayed || 3;
+        this._card = settings.card;
+        if (settings.promptId) {
+            this._promptId = settings.promptId;
+            this.usesCustomPromptId = true;
+        }
+    }
+    get inputFailMessage() {
+        return this._inputFailMessage;
+    }
+    set inputFailMessage(message) {
+        this._inputFailMessage = message;
+    }
+    get requiredInputIds() {
+        return this._requiredInputIds;
+    }
+    set requiredInputIds(ids) {
+        this._requiredInputIds = ids;
+    }
+    get missingRequiredInputsMessage() {
+        return this._missingRequiredInputsMessage;
+    }
+    set missingRequiredInputsMessage(message) {
+        this._missingRequiredInputsMessage = message;
+    }
+    get attemptsBeforeCardRedisplayed() {
+        return this._attemptsBeforeCardRedisplayed;
+    }
+    set attemptsBeforeCardRedisplayed(attempts) {
+        this._attemptsBeforeCardRedisplayed = attempts;
+    }
+    get promptId() {
+        return this._promptId;
+    }
+    set promptId(id) {
+        this.usesCustomPromptId = true;
+        this._promptId = id;
+    }
+    get card() {
+        return this._card;
+    }
+    set card(card) {
+        this._card = card;
+    }
+    async beginDialog(dc, options) {
+        // Initialize prompt state
+        const state = dc.activeDialog.state;
+        state.options = options;
+        state.state = {};
+        // Send initial prompt
+        await this.onPrompt(dc.context, state.state, state.options, false);
+        return Dialog.EndOfTurn;
+    }
+    async onPrompt(context, state, options, isRetry) {
+        // Should use GUID for C# -- it isn't native to Node, so this keeps dependencies down
+        // Only the most recently-prompted card submission is accepted
+        this._promptId = this.usesCustomPromptId ? this._promptId : `${Math.random()}`;
+        let prompt = isRetry && options.retryPrompt ? options.retryPrompt : options.prompt;
+        // Create a prompt if user didn't pass it in through PromptOptions
+        if (!prompt || Object.keys(prompt).length === 0 || typeof (prompt) != 'object' || !prompt.attachments || prompt.attachments.length === 0) {
+            prompt = {
+                attachments: [],
+                text: typeof (prompt) === 'string' ? prompt : undefined,
+            };
+        }
+        // Use card passed in PromptOptions or if it doesn't exist, use the one passed in from the constructor
+        const card = prompt.attachments && prompt.attachments[0] ? prompt.attachments[0] : this._card;
+        this.validateIsCard(card, isRetry);
+        prompt.attachments = [this.addPromptIdToCard(card)];
+        await context.sendActivity(prompt, undefined, InputHints.ExpectingInput);
```

Comment:
This behavior is different from C# which does an odd [`Insert`](https://github.com/microsoft/BotBuilder-Samples/pull/1616/files#diff-b094f412a8ac18abb94d53300adefcc5R162) of the manipulated card back into the `prompt` Activity. Here it's removing all previous attachments and only using the modified AdaptiveCard.

This is a scenario where we should be cloning the `prompt` and sending it. The `clonedPrompt` would have the AdaptiveCard that has the promptId. The original `promptOptions` which is saved in the state store should not be modified like this.

Reply:
I actually changed that in the new PR because I noticed that. This will all change when the card is used via constructor.